nixos/modules/services/games/xonotic.nix

```diff
@@ -0,0 +1,130 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) mkEnableOption mkIf mkOption types;
+
+  cfg = config.services.xonotic;
+
+  toCfg = lib.generators.toKeyValue {
+    mkKeyValue = lib.generators.mkKeyValueDefault {
+      mkValueString = v:
+        if v == true then
+          "1"
+        else if v == false then
+          "0"
+        else if builtins.isString v then
+          ''"${v}"''
+        else
+          lib.generators.mkValueStringDefault { } v;
+    } " ";
+  };
+
+  toConfig = config: toCfg config;
+
+  writeConfig = { config, extraConfig }:
+    pkgs.writeText "server.cfg" ''
+      ${toConfig config}
+      ${extraConfig}
+    '';
+
+  serverModule = types.submodule ({ name, ... }: {
+    options = {
+      config = mkOption {
+        type = types.attrsOf (types.oneOf [ types.str types.int types.bool ]);
+        description = ''
+          server config, see
+          <link xlink:href="https://github.com/xonotic/xonotic/wiki/basic-server-configuration">https://github.com/xonotic/xonotic/wiki/basic-server-configuration</link>
+          and
+          <link xlink:href="https://xonotic.org/tools/cacs/#0a/0/">https://xonotic.org/tools/cacs/#0a/0/</link>
+        '';
+        default = {
+          hostname = name;
+          port = 26000;
+        };
+      };
+
+      extraConfig = mkOption {
+        type = types.lines;
+        description = "extra config written to server.cfg";
+        example = ''
+          duel
+        '';
+        default = "";
+      };
+
+      preStart = mkOption {
+        type = types.lines;
+        description = "command to execute before starting xonotic";
+        example = "ln -s ${./warfare.pk3} ./data/maps"; # TODO: validate on how to install a map
+        default = "";
+      };
+
+      openFirewall = (mkEnableOption "open udp port for xonotic") // { default = true; };
```

Comment:
We have a rule against opening firewall holes without explicit consent from the sysadmin.